File: dashboard/ios/AttobotDashboard/Views/Screens/OverviewScreen.swift
import SwiftUI

/// Overview dashboard. Polls /api/overview + /api/workflows?status=failed&limit=8.
struct OverviewScreen: View {
    let navigate: (Route) -> Void

    @State private var overviewRes = Resource<Overview>()
    @State private var failedRes = Resource<WorkflowList>()

    var body: some View {
        ScreenScroll {
            if overviewRes.isLoading && overviewRes.value == nil {
                Spinner()
            } else if let err = overviewRes.error, overviewRes.value == nil {
                ErrorState(message: err)
            } else if let o = overviewRes.value {
                content(o: o)
            } else {
                Spinner()
            }
        }
        .navigationTitle("Overview")
        .navigationBarTitleDisplayMode(.inline)
        .task {
            overviewRes.start(
                load: { try await APIClient.get("/api/overview") },
                pollInterval: .seconds(5)
            )
            failedRes.start(
                load: {
                    try await APIClient.get(APIClient.buildPath(
                        "/api/workflows",
                        params: [("status", "failed"), ("limit", "8")]
                    ))
                },
                pollInterval: .seconds(5)
            )
        }
        .onDisappear {
            overviewRes.cancel()
            failedRes.cancel()
        }
    }

    @ViewBuilder
    private func content(o: Overview) -> some View {
        let m = o.metrics ?? [:]
        Text("Overview")
            .font(.system(size: 20, weight: .bold))
            .foregroundStyle(Theme.text)
            .padding(.bottom, 12)

        let metrics: [(String, String)] = [
            ("total instances", metricText(m["total_instances"])),
            ("running", metricText(m["running_instances"])),
            ("completed", metricText(m["completed_instances"])),
            ("failed", metricText(m["failed_instances"])),
            ("total executions", metricText(m["total_executions"])),
            ("total events", metricText(m["total_events"])),
        ]

        LazyVGrid(columns: [GridItem(.flexible(), spacing: 12), GridItem(.flexible(), spacing: 12)], spacing: 12) {
            ForEach(metrics, id: \.0) { label, val in
                MetricTile(value: val, label: label)
            }
        }
        .padding(.bottom, 8)

        Card(title: "pg_durable worker") {
            if let w = o.worker {
                workerBody(w: w)
            } else {
                EmptyState("Worker liveness unavailable.")
            }
        }

        Card(title: "By status") {
            if let bs = o.byStatus, !bs.isEmpty {
                VStack(spacing: 0) {
                    ForEach(Array(bs.enumerated()), id: \.offset) { _, r in
                        Button {
                            navigate(.workflows(status: r.status))
                        } label: {
                            HStack(spacing: 10) {
                                StatusBadge(r.status)
                                Spacer()
                                Text(r.count)
                                    .font(.system(size: 14))
                                    .foregroundStyle(Theme.muted)
                            }
                            .contentShape(Rectangle())
                            .padding(.vertical, 6)
                        }
                        .buttonStyle(.plain)
                    }
                }
            } else {
                EmptyState("No instances.")
            }
        }

        Card(title: "By type") {
            if let bt = o.byType, !bt.isEmpty {
                VStack(spacing: 0) {
                    ForEach(Array(bt.enumerated()), id: \.offset) { _, r in
                        Button {
                            navigate(.workflows(type: r.type))
                        } label: {
                            HStack(spacing: 10) {
                                TypePill(r.type ?? "")
                                Spacer()
                                Text(r.count)
                                    .font(.system(size: 14))
                                    .foregroundStyle(Theme.muted)
                            }
                            .contentShape(Rectangle())
                            .padding(.vertical, 6)
                        }
                        .buttonStyle(.plain)
                    }
                }
            } else {
                EmptyState("No instances.")
            }
        }

        Card(title: "Agents") {
            if let agents = o.agents, !agents.isEmpty {
                VStack(spacing: 0) {
                    ForEach(Array(agents.enumerated()), id: \.offset) { _, a in
                        Button {
                            navigate(.messages(agentId: String(a.id)))
                        } label: {
                            HStack(spacing: 10) {
                                Circle()
                                    .fill(a.enabled ? Theme.ok : Theme.err)
                                    .frame(width: 9, height: 9)
                                Text(a.slug)
                                    .font(.system(size: 14))
                                    .foregroundStyle(Theme.text)
                                Spacer()
                            }
                            .contentShape(Rectangle())
                            .padding(.vertical, 6)
                        }
                        .buttonStyle(.plain)
                    }
                }
            } else {
                EmptyState("No agents.")
            }
        }

        Card(title: "Recent failed workflows") {
            if failedRes.isLoading && failedRes.value == nil {
                Spinner()
            } else {
                DataTable(
                    columns: [
                        DataTableColumn<WorkflowRow>(id: "id", title: "id", width: 140) { r in
                            Text(r.id)
                                .font(.system(size: 12, design: .monospaced))
                                .foregroundStyle(Theme.accent)
                                .lineLimit(1)
                        },
                        DataTableColumn<WorkflowRow>(id: "label", title: "label") { r in
                            Text(r.label).font(.system(size: 13)).foregroundStyle(Theme.text).lineLimit(1)
                        },
                        DataTableColumn<WorkflowRow>(id: "type", title: "type") { r in
                            TypePill(r.type)
                        },
                        DataTableColumn<WorkflowRow>(id: "status", title: "status") { r in
                            StatusBadge(r.status)
                        },
                        DataTableColumn<WorkflowRow>(id: "updated", title: "updated") { r in
                            Text(Format.timeAgo(r.updatedAt)).font(.system(size: 13)).foregroundStyle(Theme.text)
                        },
                    ],
                    rows: failedRes.value?.rows ?? [],
                    onRow: { r in navigate(.workflowDetail(id: r.id)) }
                )
            }
        }
    }

    @ViewBuilder
    private func workerBody(w: Overview.Worker) -> some View {
        let alive = (w.ageSeconds ?? 999) < 15
        KeyValue(items: [
            ("status", AnyView(
                HStack(spacing: 8) {
                    Circle()
                        .fill(alive ? Theme.ok : Theme.err)
                        .frame(width: 9, height: 9)
                    Text(alive ? "alive" : "stale / down")
                        .font(.system(size: 13))
                        .foregroundStyle(Theme.text)
                }
            )),
            ("last heartbeat", AnyView(
                Text(w.ageSeconds != nil ? String(format: "%.1fs ago", w.ageSeconds!) : "—")
                    .font(.system(size: 13))
                    .foregroundStyle(Theme.text)
            )),
            ("started", AnyView(
                Text(Format.formatDateTime(w.startedAt))
                    .font(.system(size: 13))
                    .foregroundStyle(Theme.text)
            )),
        ])
    }

    private func metricText(_ v: JSONValue?) -> String {
        guard let v, !v.isNull else { return "—" }
        if let s = v.string { return s }
        if let n = v.numberValue {
            if n == n.rounded() && abs(n) < 1e15 { return String(Int64(n)) }
            return String(n)
        }
        if let b = v.boolValue { return String(b) }
        return "—"
    }
}

private struct MetricTile: View {
    let value: String
    let label: String
    var body: some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(value)
                .font(.system(size: 24, weight: .bold))
                .foregroundStyle(Theme.text)
            Text(label)
                .font(.system(size: 12))
                .foregroundStyle(Theme.muted)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(14)
        .background(Theme.panel)
        .overlay(
            RoundedRectangle(cornerRadius: 10)
                .stroke(Theme.border, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: 10))
    }
}

File: dashboard/android/app/src/main/java/com/OpenWorld/dashboard/ui/screens/OverviewScreen.kt
package com.OpenWorld.dashboard.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import androidx.navigation.NavHostController
import kotlinx.serialization.json.JsonElement
import com.OpenWorld.dashboard.core.Overview
import com.OpenWorld.dashboard.core.WorkflowRow
import com.OpenWorld.dashboard.core.fieldOr
import com.OpenWorld.dashboard.core.formatDateTime
import com.OpenWorld.dashboard.core.timeAgo
import com.OpenWorld.dashboard.state.OverviewViewModel
import com.OpenWorld.dashboard.state.UiState
import com.OpenWorld.dashboard.ui.components.OpenWorldCard
import com.OpenWorld.dashboard.ui.components.DataTable
import com.OpenWorld.dashboard.ui.components.EmptyState
import com.OpenWorld.dashboard.ui.components.ErrorState
import com.OpenWorld.dashboard.ui.components.KeyValue
import com.OpenWorld.dashboard.ui.components.LoadingView
import com.OpenWorld.dashboard.ui.components.PullRefreshScreen
import com.OpenWorld.dashboard.ui.components.StatusBadge
import com.OpenWorld.dashboard.ui.components.TableColumn
import com.OpenWorld.dashboard.ui.components.TypePill
import com.OpenWorld.dashboard.ui.nav.Routes
import com.OpenWorld.dashboard.ui.theme.accent
import com.OpenWorld.dashboard.ui.theme.border
import com.OpenWorld.dashboard.ui.theme.err
import com.OpenWorld.dashboard.ui.theme.muted
import com.OpenWorld.dashboard.ui.theme.ok
import com.OpenWorld.dashboard.ui.theme.panel
import com.OpenWorld.dashboard.ui.theme.text

@Composable
fun OverviewScreen(navController: NavHostController) {
    val vm: OverviewViewModel = viewModel()
    val state by vm.state.collectAsStateWithLifecycle()
    val s = state
    val refreshing = s is UiState.Ready && s.refreshing

    PullRefreshScreen(refreshing = refreshing, onRefresh = vm::refresh) {
        Text(
            "Overview",
            color = text,
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold,
            modifier = Modifier.padding(bottom = 12.dp),
        )
        when (s) {
            is UiState.Loading -> LoadingView()
            is UiState.Error -> ErrorState(s.message)
            is UiState.Ready -> OverviewBody(s.data.overview, s.data.failed, navController)
        }
    }
}

@Composable
private fun OverviewBody(o: Overview, failed: List<WorkflowRow>, navController: NavHostController) {
    MetricGrid(o.metrics)

    OpenWorldCard(title = "pg_durable worker") {
        val w = o.worker
        if (w != null) {
            val alive = (w.ageSeconds ?: 999.0) < 15.0
            KeyValue(
                items = listOf(
                    "status" to {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Dot(if (alive) ok else err)
                            Text(
                                if (alive) "alive" else "stale / down",
                                color = text,
                                fontSize = 13.sp,
                                modifier = Modifier.padding(start = 8.dp),
                            )
                        }
                    },
                    "last heartbeat" to {
                        Text(
                            if (w.ageSeconds != null) "%.1fs ago".format(w.ageSeconds) else "—",
                            color = text,
                            fontSize = 13.sp,
                        )
                    },
                    "started" to {
                        Text(formatDateTime(w.startedAt), color = text, fontSize = 13.sp)
                    },
                ),
            )
        } else {
            EmptyState("Worker liveness unavailable.")
        }
    }

    OpenWorldCard(title = "By status") {
        if (o.byStatus.isEmpty()) {
            EmptyState("No instances.")
        } else {
            Column {
                o.byStatus.forEach { r ->
                    CountRow(
                        onClick = { navController.navigate(Routes.workflows(status = r.status)) },
                    ) {
                        StatusBadge(r.status)
                        Text(
                            r.count,
                            color = muted,
                            fontSize = 14.sp,
                            modifier = Modifier.padding(start = 10.dp),
                        )
                    }
                }
            }
        }
    }

    OpenWorldCard(title = "By type") {
        if (o.byType.isEmpty()) {
            EmptyState("No instances.")
        } else {
            Column {
                o.byType.forEach { r ->
                    CountRow(
                        onClick = { navController.navigate(Routes.workflows(type = r.type)) },
                    ) {
                        TypePill(r.type)
                        Text(
                            r.count,
                            color = muted,
                            fontSize = 14.sp,
                            modifier = Modifier.padding(start = 10.dp),
                        )
                    }
                }
            }
        }
    }

    OpenWorldCard(title = "Agents") {
        if (o.agents.isEmpty()) {
            EmptyState("No agents.")
        } else {
            Column {
                o.agents.forEach { a ->
                    CountRow(
                        onClick = { navController.navigate(Routes.withAgent("messages", a.id.toString())) },
                    ) {
                        Dot(if (a.enabled) ok else err)
                        Text(
                            a.slug,
                            color = text,
                            fontSize = 14.sp,
                            modifier = Modifier
                                .weight(1f)
                                .padding(start = 10.dp),
                        )
                    }
                }
            }
        }
    }

    OpenWorldCard(title = "Recent failed workflows") {
        DataTable(
            rows = failed,
            onRow = { navController.navigate(Routes.workflowDetail(it.id)) },
            columns = failedColumns(),
        )
    }
}

private fun failedColumns(): List<TableColumn<WorkflowRow>> = listOf(
    TableColumn(header = "id", width = 140.dp) {
        Text(it.id, color = accent, fontFamily = FontFamily.Monospace, fontSize = 12.sp, maxLines = 1)
    },
    TableColumn(header = "label", width = 200.dp) {
        Text(it.label, color = text, fontSize = 13.sp, maxLines = 1)
    },
    TableColumn(header = "type", width = 80.dp) {
        TypePill(it.type)
    },
    TableColumn(header = "status", width = 110.dp) {
        StatusBadge(it.status)
    },
    TableColumn(header = "updated", width = 90.dp) {
        Text(timeAgo(it.updatedAt), color = text, fontSize = 13.sp)
    },
)

@Composable
private fun MetricGrid(metrics: Map<String, JsonElement>?) {
    val tiles = listOf(
        "total instances" to (metrics?.fieldOr("total_instances", "—") ?: "—"),
        "running" to (metrics?.fieldOr("running_instances", "—") ?: "—"),
        "completed" to (metrics?.fieldOr("completed_instances", "—") ?: "—"),
        "failed" to (metrics?.fieldOr("failed_instances", "—") ?: "—"),
        "total executions" to (metrics?.fieldOr("total_executions", "—") ?: "—"),
        "total events" to (metrics?.fieldOr("total_events", "—") ?: "—"),
    )
    tiles.chunked(2).forEach { pair ->
        Row(
            Modifier
                .fillMaxWidth()
                .padding(bottom = 12.dp),
            horizontalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            pair.forEach { (label, value) ->
                MetricTile(label = label, value = value, modifier = Modifier.weight(1f))
            }
        }
    }
}

@Composable
private fun MetricTile(label: String, value: String, modifier: Modifier = Modifier) {
    androidx.compose.material3.Surface(
        color = panel,
        shape = RoundedCornerShape(10.dp),
        border = androidx.compose.foundation.BorderStroke(1.dp, border),
        modifier = modifier,
    ) {
        Column(Modifier.padding(14.dp)) {
            Text(value, color = text, fontSize = 24.sp, fontWeight = FontWeight.Bold)
            Text(
                label,
                color = muted,
                fontSize = 12.sp,
                modifier = Modifier.padding(top = 2.dp),
            )
        }
    }
}

@Composable
private fun CountRow(onClick: () -> Unit, content: @Composable RowScope.() -> Unit) {
    Row(
        Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(vertical = 6.dp),
        verticalAlignment = Alignment.CenterVertically,
        content = content,
    )
}

@Composable
private fun Dot(color: Color) {
    Box(
        Modifier
            .size(9.dp)
            .clip(CircleShape)
            .background(color),
    )
}
